File: ui/trading-bot-ui/src/services/brokerProfileService.js
// services/brokerProfileService.js
import api from "./api";

class BrokerProfileService {
  constructor() {
    this.baseURL = "/v1/broker-profile";
  }

  /**
   * Get user profile for a specific broker
   * @param {string} brokerName - Broker name (upstox, angel, dhan)
   * @returns {Promise<Object>} Broker profile data
   */
  async getBrokerProfile(brokerName) {
    try {
      const response = await api.get(
        `${this.baseURL}/profile/${brokerName.toLowerCase()}`
      );
      return response.data;
    } catch (error) {
      console.error(`Error fetching ${brokerName} profile:`, error);
      throw this.handleError(error, `Failed to fetch ${brokerName} profile`);
    }
  }

  /**
   * Get funds and margin for a specific broker
   * @param {string} brokerName - Broker name
   * @param {string} segment - Optional segment (SEC for equity, COM for commodity)
   * @returns {Promise<Object>} Funds and margin data
   */
  async getBrokerFunds(brokerName, segment = null) {
    try {
      const params = segment ? { segment } : {};
      const response = await api.get(
        `${this.baseURL}/funds/${brokerName.toLowerCase()}`,
        { params }
      );
      return response.data;
    } catch (error) {
      console.error(`Error fetching ${brokerName} funds:`, error);
      throw this.handleError(error, `Failed to fetch ${brokerName} funds`);
    }
  }

  /**
   * Get profiles for all active brokers
   * @returns {Promise<Object>} All broker profiles
   */
  async getAllBrokerProfiles() {
    try {
      const response = await api.get(`${this.baseURL}/profile/all`);
      return response.data;
    } catch (error) {
      console.error("Error fetching all broker profiles:", error);
      throw this.handleError(error, "Failed to fetch broker profiles");
    }
  }

  /**
   * Get combined funds summary across all brokers
   * @returns {Promise<Object>} Combined funds summary
   */
  async getCombinedFundsSummary() {
    try {
      const response = await api.get(`${this.baseURL}/funds-summary`);
      return response.data;
    } catch (error) {
      console.error("Error fetching combined funds summary:", error);
      throw this.handleError(error, "Failed to fetch funds summary");
    }
  }

  /**
   * Get list of supported brokers
   * @returns {Promise<Object>} Supported brokers info
   */
  async getSupportedBrokers() {
    try {
      const response = await api.get(`${this.baseURL}/supported-brokers`);
      return response.data;
    } catch (error) {
      console.error("Error fetching supported brokers:", error);
      throw this.handleError(error, "Failed to fetch supported brokers");
    }
  }

  /**
   * Get Upstox profile (backward compatibility)
   * @returns {Promise<Object>} Upstox profile
   */
  async getUpstoxProfile() {
    return this.getBrokerProfile("upstox");
  }

  /**
   * Get Upstox funds (backward compatibility)
   * @param {string} segment - Optional segment
   * @returns {Promise<Object>} Upstox funds
   */
  async getUpstoxFunds(segment = null) {
    return this.getBrokerFunds("upstox", segment);
  }

  /**
   * Format profile data for display
   * @param {Object} profileData - Raw profile data
   * @returns {Object} Formatted profile data
   */
  formatProfileData(profileData) {
    if (!profileData?.data?.data) return null;

    const data = profileData.data.data;
    return {
      email: data.email || "N/A",
      userId: data.user_id || "N/A",
      userName: data.user_name || "N/A",
      broker: data.broker || "N/A",
      userType: data.user_type || "individual",
      isActive: data.is_active || false,
      exchanges: data.exchanges || [],
      products: data.products || [],
      orderTypes: data.order_types || [],
      poa: data.poa || false,
      ddpi: data.ddpi || false,
      lastUpdated: profileData.last_updated,
    };
  }

  /**
   * Format funds data for display
   * @param {Object} fundsData - Raw funds data
   * @returns {Object} Formatted funds data
   */
  formatFundsData(fundsData) {
    if (!fundsData?.data?.data) return null;

    const data = fundsData.data.data;
    const equity = data.equity || {};
    const commodity = data.commodity || {};

    return {
      equity: {
        availableMargin: equity.available_margin || 0,
        usedMargin: equity.used_margin || 0,
        payinAmount: equity.payin_amount || 0,
        spanMargin: equity.span_margin || 0,
        adhocMargin: equity.adhoc_margin || 0,
        notionalCash: equity.notional_cash || 0,
        exposureMargin: equity.exposure_margin || 0,
        utilization:
          equity.available_margin > 0
            ? ((equity.used_margin / equity.available_margin) * 100).toFixed(2)
            : 0,
      },
      commodity: {
        availableMargin: commodity.available_margin || 0,
        usedMargin: commodity.used_margin || 0,
        payinAmount: commodity.payin_amount || 0,
        spanMargin: commodity.span_margin || 0,
        adhocMargin: commodity.adhoc_margin || 0,
        notionalCash: commodity.notional_cash || 0,
        exposureMargin: commodity.exposure_margin || 0,
        utilization:
          commodity.available_margin > 0
            ? (
                (commodity.used_margin / commodity.available_margin) *
                100
              ).toFixed(2)
            : 0,
      },
      segment: fundsData.segment,
      lastUpdated: fundsData.last_updated,
    };
  }

  /**
   * Format combined funds summary
   * @param {Object} summaryData - Raw summary data
   * @returns {Object} Formatted summary
   */
  formatCombinedSummary(summaryData) {
    if (!summaryData?.summary) return null;

    const summary = summaryData.summary;
    return {
      totalAvailable: summary.total_available_margin || 0,
      totalUsed: summary.total_used_margin || 0,
      utilization: summary.utilization_percentage || 0,
      brokerFunds: summaryData.broker_funds || {},
      lastUpdated: summaryData.last_updated,
    };
  }

  /**
   * Handle API errors consistently
   * @param {Object} error - Error object
   * @param {string} defaultMessage - Default error message
   * @returns {Error} Formatted error
   */
  handleError(error, defaultMessage) {
    if (error.response) {
      const status = error.response.status;
      const detail = error.response.data?.detail || defaultMessage;

      switch (status) {
        case 401:
          return new Error("Authentication required. Please login again.");
        case 404:
          return new Error(
            "Broker configuration not found. Please configure your broker."
          );
        case 400:
          return new Error(detail);
        case 500:
          return new Error("Server error. Please try again later.");
        default:
          return new Error(detail);
      }
    }

    return new Error(error.message || defaultMessage);
  }

  /**
   * Check if broker profile service is available
   * @param {string} brokerName - Broker name
   * @returns {Promise<boolean>} Service availability
   */
  async isBrokerSupported(brokerName) {
    try {
      const supportedBrokers = await this.getSupportedBrokers();
      const broker =
        supportedBrokers.supported_brokers?.[brokerName.toLowerCase()];
      return broker?.profile_supported || false;
    } catch (error) {
      console.error("Error checking broker support:", error);
      return false;
    }
  }

  /**
   * Cache keys for local storage
   */
  getCacheKey(type, brokerName = "", segment = "") {
    return `broker_${type}_${brokerName}_${segment}`.replace(/__+/g, "_");
  }

  /**
   * Cache profile data temporarily (5 minutes)
   * @param {string} brokerName - Broker name
   * @param {Object} data - Profile data
   */
  cacheProfile(brokerName, data) {
    const cacheKey = this.getCacheKey("profile", brokerName);
    const cacheData = {
      data,
      timestamp: Date.now(),
      expiry: Date.now() + 5 * 60 * 1000, // 5 minutes
    };

    try {
      localStorage.setItem(cacheKey, JSON.stringify(cacheData));
    } catch (error) {
      console.warn("Failed to cache profile data:", error);
    }
  }

  /**
   * Get cached profile data if valid
   * @param {string} brokerName - Broker name
   * @returns {Object|null} Cached data or null
   */
  getCachedProfile(brokerName) {
    const cacheKey = this.getCacheKey("profile", brokerName);

    try {
      const cachedData = localStorage.getItem(cacheKey);
      if (!cachedData) return null;

      const parsed = JSON.parse(cachedData);
      if (Date.now() > parsed.expiry) {
        localStorage.removeItem(cacheKey);
        return null;
      }

      return parsed.data;
    } catch (error) {
      console.warn("Failed to get cached profile data:", error);
      return null;
    }
  }
}

// Export singleton instance
const brokerProfileService = new BrokerProfileService();
export default brokerProfileService;

File: services/broker_profile_service.py
"""
Broker Profile Service
Unified service to handle broker profile and funds information across different brokers
"""

import logging
from typing import Dict, Any, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from fastapi import HTTPException

from database.models import BrokerConfig, User
from services.upstox_service import get_upstox_user_profile, get_upstox_funds_and_margin

logger = logging.getLogger(__name__)


class BrokerProfileService:
    """
    Unified service for broker profile and funds management
    Supports multiple brokers with a common interface
    """

    def __init__(self, db: Session):
        self.db = db

    def get_user_broker_profile(self, user_id: int, broker_name: str) -> Dict[str, Any]:
        """
        Get user profile information for a specific broker
        
        Args:
            user_id: User ID
            broker_name: Broker name (upstox, angel, dhan, etc.)
            
        Returns:
            Dict containing profile information
        """
        try:
            # Get broker configuration
            broker_config = self._get_active_broker_config(user_id, broker_name)
            
            # Route to appropriate broker service
            if broker_name.lower() == "upstox":
                return self._get_upstox_profile(broker_config)
            elif broker_name.lower() == "angel":
                return self._get_angel_profile(broker_config)
            elif broker_name.lower() == "dhan":
                return self._get_dhan_profile(broker_config)
            else:
                raise HTTPException(
                    status_code=400,
                    detail=f"Profile service not implemented for broker: {broker_name}"
                )

        except Exception as e:
            logger.error(f"Error fetching profile for user {user_id}, broker {broker_name}: {e}")
            raise

    def get_user_funds_and_margin(
        self, 
        user_id: int, 
        broker_name: str, 
        segment: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Get user funds and margin information for a specific broker
        
        Args:
            user_id: User ID
            broker_name: Broker name
            segment: Market segment (if applicable)
            
        Returns:
            Dict containing funds and margin information
        """
        try:
            # Get broker configuration
            broker_config = self._get_active_broker_config(user_id, broker_name)
            
            # Route to appropriate broker service
            if broker_name.lower() == "upstox":
                return self._get_upstox_funds(broker_config, segment)
            elif broker_name.lower() == "angel":
                return self._get_angel_funds(broker_config, segment)
            elif broker_name.lower() == "dhan":
                return self._get_dhan_funds(broker_config, segment)
            else:
                raise HTTPException(
                    status_code=400,
                    detail=f"Funds service not implemented for broker: {broker_name}"
                )

        except Exception as e:
            logger.error(f"Error fetching funds for user {user_id}, broker {broker_name}: {e}")
            raise

    def get_all_user_broker_profiles(self, user_id: int) -> Dict[str, Any]:
        """
        Get profile information for all active brokers of a user
        
        Args:
            user_id: User ID
            
        Returns:
            Dict with broker profiles
        """
        try:
            # Get all active broker configurations for user
            broker_configs = self.db.query(BrokerConfig).filter(
                BrokerConfig.user_id == user_id,
                BrokerConfig.is_active == True
            ).all()

            if not broker_configs:
                return {"profiles": {}, "message": "No active broker configurations found"}

            profiles = {}
            for config in broker_configs:
                try:
                    profile = self.get_user_broker_profile(user_id, config.broker_name)
                    profiles[config.broker_name.lower()] = {
                        "profile": profile,
                        "config_id": config.id,
                        "last_updated": datetime.now().isoformat()
                    }
                except Exception as e:
                    logger.warning(f"Failed to fetch profile for {config.broker_name}: {e}")
                    profiles[config.broker_name.lower()] = {
                        "error": str(e),
                        "config_id": config.id
                    }

            return {
                "profiles": profiles,
                "total_brokers": len(broker_configs),
                "successful_fetches": len([p for p in profiles.values() if "profile" in p])
            }

        except Exception as e:
            logger.error(f"Error fetching all profiles for user {user_id}: {e}")
            raise

    def get_combined_funds_summary(self, user_id: int) -> Dict[str, Any]:
        """
        Get combined funds summary across all active brokers
        
        Args:
            user_id: User ID
            
        Returns:
            Combined funds summary
        """
        try:
            logger.info(f"🔍 DEBUG: get_combined_funds_summary called for user_id: {user_id}")
            logger.info(f"🔍 DEBUG: Database session: {self.db}")
            logger.info(f"🔍 DEBUG: Service instance: {self}")
            logger.info(f"🔍 DEBUG: Method being called: get_combined_funds_summary")
            
            # TEMP DEBUG: Check what this method actually does
            import inspect
            logger.info(f"🔍 DEBUG: Method source file: {inspect.getfile(self.get_combined_funds_summary)}")
            
            
            broker_configs = self.db.query(BrokerConfig).filter(
                BrokerConfig.user_id == user_id,
                BrokerConfig.is_active == True
            ).all()

            if not broker_configs:
                return {"funds": {}, "message": "No active broker configurations found"}

            funds_data = {}
            total_available = 0
            total_used = 0

            for config in broker_configs:
                try:
                    funds = self.get_user_funds_and_margin(user_id, config.broker_name)
                    funds_data[config.broker_name.lower()] = funds
                    
                    # Extract and sum available/used margins (broker-specific logic)
                    if config.broker_name.lower() == "upstox" and "data" in funds:
                        equity_data = funds["data"].get("equity", {})
                        total_available += equity_data.get("available_margin", 0)
                        total_used += equity_data.get("used_margin", 0)

                except Exception as e:
                    logger.warning(f"Failed to fetch funds for {config.broker_name}: {e}")
                    funds_data[config.broker_name.lower()] = {"error": str(e)}

            return {
                "broker_funds": funds_data,
                "summary": {
                    "total_available_margin": total_available,
                    "total_used_margin": total_used,
                    "utilization_percentage": (total_used / total_available * 100) if total_available > 0 else 0
                },
                "last_updated": datetime.now().isoformat()
            }

        except Exception as e:
            logger.error(f"Error fetching combined funds for user {user_id}: {e}")
            raise

    def _get_active_broker_config(self, user_id: int, broker_name: str) -> BrokerConfig:
        """Get active broker configuration for user"""
        config = self.db.query(BrokerConfig).filter(
            BrokerConfig.user_id == user_id,
            BrokerConfig.broker_name.ilike(broker_name),
            BrokerConfig.is_active == True
        ).first()
        
        if not config:
            raise HTTPException(
                status_code=404,
                detail=f"{broker_name} broker configuration not found or inactive"
            )
        
        if not config.access_token:
            raise HTTPException(
                status_code=400,
                detail=f"{broker_name} access token not available. Please re-authenticate."
            )
        
        # Check token expiry
        if config.access_token_expiry and config.access_token_expiry <= datetime.now():
            raise HTTPException(
                status_code=401,
                detail=f"{broker_name} access token expired. Please re-authenticate."
            )
        
        return config

    def _get_upstox_profile(self, config: BrokerConfig) -> Dict[str, Any]:
        """Get Upstox user profile and store in database"""
        profile_data = get_upstox_user_profile(config.access_token)
        
        # Update the broker config with profile data
        config.update_profile_data(profile_data)
        self.db.commit()
        
        return {
            "broker": "upstox",
            "data": profile_data,
            "config_id": config.id
        }

    def _get_upstox_funds(self, config: BrokerConfig, segment: Optional[str]) -> Dict[str, Any]:
        """Get Upstox funds and margin and store in database"""
        funds_data = get_upstox_funds_and_margin(config.access_token, segment)
        
        # Update the broker config with funds data
        config.update_funds_data(funds_data)
        self.db.commit()
        
        return {
            "broker": "upstox",
            "data": funds_data,
            "segment": segment,
            "config_id": config.id
        }

    def _get_angel_profile(self, config: BrokerConfig) -> Dict[str, Any]:
        """Get Angel One profile - To be implemented"""
        raise HTTPException(
            status_code=501,
            detail="Angel One profile service not yet implemented"
        )

    def _get_angel_funds(self, config: BrokerConfig, segment: Optional[str]) -> Dict[str, Any]:
        """Get Angel One funds - To be implemented"""
        raise HTTPException(
            status_code=501,
            detail="Angel One funds service not yet implemented"
        )

    def _get_dhan_profile(self, config: BrokerConfig) -> Dict[str, Any]:
        """Get Dhan profile - To be implemented"""
        raise HTTPException(
            status_code=501,
            detail="Dhan profile service not yet implemented"
        )

    def _get_dhan_funds(self, config: BrokerConfig, segment: Optional[str]) -> Dict[str, Any]:
        """Get Dhan funds - To be implemented"""
        raise HTTPException(
            status_code=501,
            detail="Dhan funds service not yet implemented"
        )